File: src/leetcode/2026/03/Solution697.java
import java.util.HashMap;
import java.util.Map;

public class Solution697 {
    public int findShortestSubArray(int[] nums) {
        int degree = 0;
        int minLength = Integer.MAX_VALUE;
        Map<Integer, Integer> countMap = new HashMap<>();
        Map<Integer, Integer> firstIndexMap = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            int num = nums[i];
            countMap.put(num, countMap.getOrDefault(num, 0) + 1);

            if (!firstIndexMap.containsKey(num)) {
                firstIndexMap.put(num, i);
            }

            int currentCount = countMap.get(num);
            if (currentCount > degree) {
                degree = currentCount;
                minLength = i - firstIndexMap.get(num) + 1;
            } else if (currentCount == degree) {
                minLength = Math.min(minLength, i - firstIndexMap.get(num) + 1);
            }
        }

        return minLength;
    }

    public static void main(String[] args) {
        Solution697 solution = new Solution697();

        int[] nums = { 1, 2, 2, 3, 1 };
        int result = solution.findShortestSubArray(nums);
        System.out.println(result); // Output: 2

        nums = new int[] { 1, 2, 2, 3, 1, 4, 2 };
        result = solution.findShortestSubArray(nums);
        System.out.println(result); // Output: 6
    }
}

File: src/leetcode/2026/03/Solution697.cpp
#include <unordered_map>
#include <vector>
// #include <algorithm>
#include <cstdio>

using namespace std;

class Solution {
public:
    int findShortestSubArray(vector<int>& nums) {
        int degree = 0, minLength = nums.size();
        unordered_map<int, int> countMap;
        unordered_map<int, int> firstIndexMap;

        for (int i=0; i<nums.size(); i++) {
            int num = nums[i];
            countMap[num]++;
            if (firstIndexMap.find(num) == firstIndexMap.end()) {
                firstIndexMap[num] = i;
            }
            if (countMap[num] > degree) {
                degree = countMap[num];
                minLength = i - firstIndexMap[num] + 1;
            } else if (countMap[num] == degree) {
                minLength = min(minLength, i - firstIndexMap[num] + 1);
            }
        }
        return minLength;
    }
};

int main() {
    Solution sol;

    vector<int> nums = {1, 2, 2, 3, 1};
    int ans = sol.findShortestSubArray(nums);
    printf("%d\n", ans);

    nums = {1, 2, 2, 3, 1, 4, 2};
    ans = sol.findShortestSubArray(nums);
    printf("%d\n", ans);
}